src/overlays/actors/ovl_Bg_Ddan_Jd/z_bg_ddan_jd.c

```diff
@@ -31,17 +32,150 @@ const ActorInit Bg_Ddan_Jd_InitVars = {
 };
-*/
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Bg_Ddan_Jd/BgDdanJd_Init.s")
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Bg_Ddan_Jd/BgDdanJd_Destroy.s")
+static InitChainEntry sInitChain[] = {
+    ICHAIN_VEC3F_DIV1000(scale, 100, ICHAIN_STOP),
+};
+
+#define STATE_GO_BOTTOM 0
+#define STATE_GO_MIDDLE_FROM_BOTTOM 1
+#define STATE_GO_MIDDLE_FROM_TOP 2
+#define STATE_GO_TOP 3
+
+#define MOVE_HEIGHT_MIDDLE 140.0f
+#define MOVE_HEIGHT_TOP 700.0f
+
+#define IDLE_FRAMES 100
+
+// Since ySpeed also determines whether the platform is the shortcut to the top of the dungeon, these must be
+// different values to work correctly
+#define DEFAULT_Y_SPEED 1
+#define SHORTCUT_Y_SPEED 5
+
+void BgDdanJd_Init(Actor* thisx, GlobalContext* globalCtx) {
+    BgDdanJd* this = THIS;
+    s32 pad;
+    ColHeader* colHeader = NULL;
+
+    Actor_ProcessInitChain(thisx, sInitChain);
+    DynaPolyInfo_SetActorMove(thisx, DPM_PLAYER);
+    DynaPolyInfo_Alloc(&D_06003CE0, &colHeader);
+    this->dyna.dynaPolyId = DynaPolyInfo_RegisterActor(globalCtx, &globalCtx->colCtx.dyna, thisx, colHeader);
+    this->idleTimer = IDLE_FRAMES;
+    this->state = STATE_GO_BOTTOM;
+
+    // @bug: Missing check for actor.params < 0x40. This will cause inconsistent behavior if params >= 0x40 and the
+    // bound switch state is turned on while in the same room, as the shortcut behavior won't become enabled until the
+    // actor is reloaded.
+    if (Flags_GetSwitch(globalCtx, this->dyna.actor.params)) {
+        this->ySpeed = SHORTCUT_Y_SPEED;
+    } else {
+        this->ySpeed = DEFAULT_Y_SPEED;
+    }
+    this->actionFunc = BgDdanJd_Idle;
+}
+
+void BgDdanJd_Destroy(Actor* thisx, GlobalContext* globalCtx) {
+    BgDdanJd* this = THIS;
+
+    DynaPolyInfo_Free(globalCtx, &globalCtx->colCtx.dyna, this->dyna.dynaPolyId);
+}
+
+void BgDdanJd_Idle(BgDdanJd* this, GlobalContext* globalCtx) {
+    if (this->idleTimer != 0) {
+        this->idleTimer--;
+    }
+
+    // if this is the platform that rises all the way to the top, and the switch state has just changed to on
+    if (this->ySpeed == DEFAULT_Y_SPEED && this->dyna.actor.params < 0x40 &&
+        Flags_GetSwitch(globalCtx, this->dyna.actor.params)) {
+        this->ySpeed = SHORTCUT_Y_SPEED;
+        this->state = STATE_GO_MIDDLE_FROM_BOTTOM;
+        this->idleTimer = 0;
+        this->dyna.actor.posRot.pos.y = this->dyna.actor.initPosRot.pos.y + MOVE_HEIGHT_MIDDLE;
+        func_800800F8(globalCtx, 0xBF4, -0x63, &this->dyna.actor, 0);
+    }
+    if (this->idleTimer == 0) {
+        this->idleTimer = IDLE_FRAMES;
+        if (this->state == STATE_GO_BOTTOM) {
+            this->state = STATE_GO_MIDDLE_FROM_BOTTOM;
+            this->targetY = this->dyna.actor.initPosRot.pos.y + MOVE_HEIGHT_MIDDLE;
+        } else if (this->state == STATE_GO_MIDDLE_FROM_BOTTOM) {
+            // If platform goes all the way to the top
+            if (this->ySpeed != DEFAULT_Y_SPEED) {
```

Comment:
Sorry to insist but I don't see how this makes sense

```c
    // if this is the platform that rises all the way to the top, and the switch state has just changed to on
    if (this->ySpeed == DEFAULT_Y_SPEED
```
and later
```c
            // If platform goes all the way to the top
            if (this->ySpeed != DEFAULT_Y_SPEED) {
```

Reply:
Then let's follow the logic.

In `BgDdanJd_Init`, your initial state is set to `STATE_GO_BOTTOM` and ySpeed set to `DEFAULT_Y_SPEED` or `SHORCUT_Y_SPEED`, depending on whether Flags_GetSwitch returns switch off or switch on respectively. If `ySpeed == SHORTCUT_Y_SPEED` at this stage, then the actor will **ALWAYS** exhibit the shortcut behavior. You can trace BgDdanJd_Idle's logic to see this to be true: 

1. `STATE_GO_BOTTOM -> STATE_GO_MIDDLE_FROM_BOTTOM`
2. `this->ySpeed != DEFAULT_Y_SPEED  /* true */ -> STATE_GO_TOP`
3. `STATE_GO_TOP -> STATE_GO_MIDDLE_FROM_TOP`
4. repeat 2 and 3 indefinitely

Now, if BgDdanJd_Init sets ySpeed to `DEFAULT_Y_SPEED`, then it will exhibit the bottom <-> middle behavior, **BUT** it can be either...

1. A platform that **ALWAYS** exhibits the bottom <-> middle behavior regardless of switch state
2. The shortcut platform, which swaps to the shortcut behavior on switch flag on.

As such, in the first comment ySpeed == DEFAULT_Y_SPEED on it's own is not enough to determine whether the platform is the shortcut platform or not; it's only by checking params that we can determine that yes, this is the platform that can go all the way to the top.

In the second comment, ySpeed != DEFAULT_Y_SPEED, so the platform is the shortcut platform because it exhibits the shortcut behavior.

Reply:
Thanks, I get it now, so the first comment should state "if this is the platform that can become a shortcut" (refering to the actor variable, not to the `ySpeed` value?), and the second "if this platform is the enabled shortcut" or something like that?